storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -96,57 +142,119 @@ func (s *commitLogSource) Read(
+		bopts        = s.opts.ResultOptions()
+		blopts       = bopts.DatabaseBlockOptions()
+		blockSize    = ns.Options().RetentionOptions().BlockSize()
+		encoderPool  = bopts.DatabaseBlockOptions().EncoderPool()
+		workerErrs   = make([]int, numConc)
 	)
-	for iter.Next() {
-		series, dp, unit, annotation := iter.Current()
 
-		// check if the series belongs to current namespace being bootstrapped
-		if !namespace.Equal(series.Namespace) {
-			continue
+	// +1 so we can use the shard number as an index throughout without constantly
+	// remembering to subtract 1 to convert to zero-based indexing
+	unmerged := make([]encodersAndRanges, highestShard+1)
+	for shard := range shardsTimeRanges {
+		unmerged[shard] = encodersAndRanges{
+			encodersBySeries: make(map[ts.Hash]encodersByTime),
+			ranges:           shardsTimeRanges[shard],
 		}
+	}
 
-		ranges, ok := shardsTimeRanges[series.Shard]
-		if !ok {
-			// Not bootstrapping this shard
+	encoderChans := make([]chan encoderArg, numConc)
+	for i := 0; i < numConc; i++ {
+		encoderChans[i] = make(chan encoderArg, encoderChanBufSize)
+	}
+
+	// Spin up numConc background go-routines to handle M3TSZ encoding. This must
+	// happen before we start reading to prevent infinitely blocking writes to
+	// the encoderChans.
+	wg := &sync.WaitGroup{}
+	for workerNum, encoderChan := range encoderChans {
+		wg.Add(1)
+		go s.startM3TSZEncodingWorker(
+			workerNum,
+			encoderChan,
+			unmerged,
+			encoderPool,
+			workerErrs,
+			blopts,
+			wg,
+		)
+	}
+
+	for iter.Next() {
+		series, dp, unit, annotation := iter.Current()
+		if !s.shouldEncodeSeries(namespace, unmerged, blockSize, series, dp) {
 			continue
 		}
 
-		blockStart := dp.Timestamp.Truncate(blockSize)
-		blockEnd := blockStart.Add(blockSize)
-		blockRange := xtime.Range{
-			Start: blockStart,
-			End:   blockEnd,
-		}
-		if !ranges.Overlaps(blockRange) {
-			// Data in this block does not match the requested ranges
-			continue
+		// Distribute work such that each encoder goroutine is responsible for
+		// approximately numShards / numConc shards. This also means that all
+		// datapoints for a given shard/series will be processed in a serialized
+		// manner.
+		encoderChans[series.Shard%uint32(numConc)] <- encoderArg{
+			series:     series,
+			dp:         dp,
+			unit:       unit,
+			annotation: annotation,
+			blockStart: dp.Timestamp.Truncate(blockSize),
 		}
+	}
+	for _, encoderChan := range encoderChans {
+		close(encoderChan)
+	}
 
-		unmergedShard, ok := unmerged[series.Shard]
-		if !ok {
-			unmergedShard = make(map[ts.Hash]encoderMap)
-			unmerged[series.Shard] = unmergedShard
+	// Block until all data has been read and encoded by the worker goroutines
+	wg.Wait()
+	s.logEncodingOutcome(workerErrs, iter)
+
+	return s.mergeShards(int(highestShard+1), bopts, blopts, encoderPool, unmerged), nil
+}
+
+func (s *commitLogSource) findHighestShard(shardsTimeRanges result.ShardTimeRanges) uint32 {
```

Comment:
per above, move towards bottom of the file

Reply:
fixed